### PR TITLE
Offline-First TWA demo (Iteration 1)

### PR DESCRIPTION
**Pull request by demianrenzulli**
Iteration 1 for the "Offline-FIrst TWA demo". See #273 for details on iterations.

### Steps:

- Run the app on the phone. Note: The first time the app is ran you can't be in airplane mode, because Android Studio would throw an error. Alternatively, to run the first test in "offline-first" state, and .apk file can be generated and installed in the phone (outside of Android Studio).
- The TWA with https://airhorner.com/ is launched.
- Close the app and go to Airplane Mode.
- Open the app: the offline fallback screen is shown.
- Clicking the "Retry" button in offline state, doesn't open the TWA, because the user is offline.
- Go back online.
- Click the "Retry" button.
- The TWA is launched.

### Implementation details:

- `OfflineFirstTWALauncherActivity` extends `com.google.androidbrowserhelper.trusted.LauncherActivity` to prevent the default launch of the TWA, by overriding `shouldLaunchImmediately()` to return `false`.
- In the `onCreate` method, ti checks for connectivity. If the user is offline, it shows an Android offline fallback screen, with a "Retry" button.
- If the user is online, the TWA is launched.

### Observations:

In this first iteration the TWA will always show the native fallback screen when it's opened in offline state. This might be good enough in most cases, but we can explore a second iteration where we either:

- Save a flag in sharedPreferences the first time the TWA is launched, assuming that the SW might be registered at that point, and the web offline fallback page can be shown. This is not 100% accurate, since the user  could have cleaned the storage after the first run, in which case the preference would be in an inconsistent state.
- Check if the SW is registered or the html fallback page is in the cache, if so, always launch the TWA in offline state, instead of showing the Android fallback screen.

cc // @andreban @b1tr0t @aranhacarlos

### CODE REVIEW

File: demos/twa-offline-first/build.gradle [hunk 1]
```diff
@@ -0,0 +1,39 @@
+plugins {
```
Comment (andreban) [inline demos/twa-offline-first/build.gradle:1]: nit: please, add license headers to the files and ensure all files finish with an empty line.

File: demos/twa-offline-first/build.gradle [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    compileOptions {
+        sourceCompatibility JavaVersion.VERSION_1_8
+        targetCompatibility JavaVersion.VERSION_1_8
+    }
+}
+
+dependencies {
+
+    implementation 'androidx.appcompat:appcompat:1.1.0'
+    implementation 'com.google.android.material:material:1.1.0'
```
Comment (andreban) [inline demos/twa-offline-first/build.gradle:33]: How much does using this add to the APK size? If the Activity is relatively simple from a UI perspective, we may get  away without adding many extra libraries.

File: demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        this.displayScreen();
+    }
+
+    @Override
+    protected boolean shouldLaunchImmediately() {
+        return false;
+    }
+
+    private void displayScreen() {
+        if(isOnline()) {
```
Comment (andreban) [inline demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java:25]: style: space between `if` and the parenthesis.
```suggestion
        if (isOnline()) {
```

File: demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        super.onCreate(savedInstanceState);
+        this.displayScreen();
+    }
+
+    @Override
+    protected boolean shouldLaunchImmediately() {
+        return false;
+    }
+
+    private void displayScreen() {
```
Comment (andreban) [inline demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java:24]: I got confused by the method name when reading the method below. What about renaming it to something like `tryLaunchTwa()`?
```suggestion
    private void tryLaunchTwa() {
```

File: demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        if(isOnline()) {
+            launchTwa();
+        } else {
+            launchCustomOfflineScreen();
+        }
+    }
+
+    private void launchCustomOfflineScreen() {
```
Comment (andreban) [inline demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java:32]: nit: Launch makes me thing we're launching an Activity, but we're not actually launching anything here, and just setting the content of the currentActivity. Maybe rename this to `renderOfflineFallback()` or something similar?
```suggestion
    private void renderOfflineFallback() {
```

**Review by andreban — COMMENTED** [review]
Dropped some notes.

**Comment by demianrenzulli** [comment]
@andreban thanks for your review. I think I worked on all your comments now.

Regarding dependencies, I cleaned up most of them. I left  `com.google.android.material:material:1.1.0`, which lets me use the `android:theme="@style/Theme.Design.NoActionBar"` theme in `AndroidManfiest.xml`, to hide the title bar in the offline fallback screen, and make it look more similar to airhorner.

That seems to be the easiest way to  do it. You can also do it programmatically, but that would add noise to the Activity code.

 Let me know your thoughts.

File: demos/twa-offline-first/build.gradle [hunk 6]
```diff
@@ -0,0 +1,50 @@
+/*
+ *    Copyright 2019 Google LLC
```
Comment (PEConn) [inline demos/twa-offline-first/build.gradle:2]: Nit: could you go through all of the copyright headers and update the year to 2021 please?

File: settings.gradle [hunk 7]
```diff
@@ -17,3 +17,4 @@ include ':demos:twa-orientation'
 include ':demos:twa-play-billing'
 include ':demos:twa-web-share-target'
 include ':demos:twa-webview-fallback'
+include ':demos:twa-offline-first'
```
Comment (PEConn) [inline settings.gradle:20]: nit: Could you keep the demos sorted in alphabetical order?

File: demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import android.os.Bundle;
+import android.view.View;
+import android.widget.Button;
+
+public class OfflineFirstTWALauncherActivity extends LauncherActivity {
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        this.tryLaunchTwa();
```
Comment (PEConn) [inline demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java:30]: nit: no need for `this`.

File: demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Override
+    protected boolean shouldLaunchImmediately() {
+        return false;
```
Comment (PEConn) [inline demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java:35]: Could you add a comment just explaining the consequences of this line? (This is a demo so we may as well make things as obvious as possible.)

File: demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            launchTwa();
+        } else {
+            renderOfflineFallback();
+        }
+    }
+
+    private void renderOfflineFallback() {
+        setContentView(R.layout.activity_offline_first_twa);
+
+        final Button retryBtn = this.findViewById(R.id.retry_btn);
```
Comment (PEConn) [inline demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java:49]: nit: We don't normally put final on local variables.

File: demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        retryBtn.setOnClickListener(new View.OnClickListener() {
+            public void onClick(View v) {
+                tryLaunchTwa();
```
Comment (PEConn) [inline demos/twa-offline-first/src/main/java/com/google/androidbrowserhelper/demos/twa_offline_first/OfflineFirstTWALauncherActivity.java:52]: If the user is offline, and they click the button when they're still offline, we're going to be repeating all of this code even though there's no need to.

Maybe tryLaunchTwa should be something like:

```
if (isOnline()) {
  launchTwa();
} else if (!mOfflineFallbackRendered) {
  mOfflineFallbackRendered = true;
  renderOfflineFallback();
}
```

Or alternatively we just change this line to:

```
if (isOnline()) launchTwa();
```

Which is probably a bit simpler.

**Review by PEConn — COMMENTED** [review]
Hey, thanks for the contribution.

I've left a few comments - they're all pretty simple things.

**Comment by demianrenzulli** [comment]
Thanks for taking the time to review this PR @PEConn! I think I went through all your comments.

Let me know if you  have any additional thoughts.

Cheers.

**Review by PEConn — APPROVED** [review]

**Review by andreban — APPROVED** [review]
LGTM too

**Pull request merged**